httpcore/_async/http11.py

```diff
@@ -281,16 +288,28 @@ def is_available(self) -> bool:
     def has_expired(self) -> bool:
         now = time.monotonic()
         keepalive_expired = self._expire_at is not None and now > self._expire_at
+        if keepalive_expired:
+            return True
 
         # If the HTTP connection is idle but the socket is readable, then the
         # only valid state is that the socket is about to return b"", indicating
         # a server-initiated disconnect.
-        server_disconnected = (
-            self._state == HTTPConnectionState.IDLE
-            and self._network_stream.get_extra_info("is_readable")
-        )
+        # Checking the readable status is relatively expensive so check it at a lower frequency.
+        if (now - self._network_stream_used_at) > self._socket_poll_interval():
+            self._network_stream_used_at = now
+            server_disconnected = (
+                self._state == HTTPConnectionState.IDLE
+                and self._network_stream.get_extra_info("is_readable")
```

Comment:
I'm still slightly worried about what this check does behind the scenes. Does the socket poll cause IO so in async code it disturbs other requests being processed as this is doing the IO in non-async code? Maybe its more fine when we reduce the amount of polling happening.

Reply:
Another option would be to use this style of approach for checking aliveness. So preemptively trying to load data from the stream into a buffer. https://github.com/encode/httpx/issues/2983#issuecomment-1843756808

Reply:
Or yet another option would be to do this similarly as in `urllib3` so we would do the socket poll check when picking up the connection from the pool and not as part of the pool maintenance loop. Like here [is_connection_dropped](https://github.com/urllib3/urllib3/blob/main/src/urllib3/connectionpool.py#L290) which uses [wait_for_read](https://github.com/urllib3/urllib3/blob/da410581b6b3df73da976b5ce5eb20a4bd030437/src/urllib3/connection.py#L260) doing the [pool](https://github.com/urllib3/urllib3/blob/da410581b6b3df73da976b5ce5eb20a4bd030437/src/urllib3/util/wait.py#L70) similarly as in `httpcore`